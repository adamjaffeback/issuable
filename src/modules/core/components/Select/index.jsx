import React, { Component } from 'react';

import { classie } from 'utils';

import './styles.scss';

const Option = ({ value, label, className, hidden }) => (
  <option
    hidden={hidden}
    className={classie(['option', className])}
    value={value}>
    {label}
  </option>
);

class Select extends Component {
  static defaultProps = {
    onChange: () => null,
    value: {},
    options: [],
    showPlaceholder: true,
    placeholder: 'Select',
  };

  handleChange = (evt) => {
    const { onChange } = this.props;
    const { target: { value, name } } = evt;
    onChange(value, name);
  };

  render() {
    const {
      value,
      options,
      name,
      className,
      optionClassName,
      showPlaceholder,
      placeholder,
    } = this.props;

    return (
      <select
        value={value || ''}
        name={name}
        onChange={this.handleChange}
        className={classie(['select', className])}>
        {showPlaceholder &&
          <Option
            hidden={true}
            label={placeholder}
            className={optionClassName}
            value="" />}

        {options.map((opt, idx) => (
          <Option
            value={opt.value}
            key={`${name}-option-${idx}`}
            label={opt.label}
            className={optionClassName} />
        ))}
      </select>
    );
  }
}

export default Select;
