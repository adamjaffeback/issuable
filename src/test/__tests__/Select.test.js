import React from 'react';
import { shallow, render } from 'enzyme';
import renderer from 'react-test-renderer';
import Select from '../../modules/core/components/Select';

const options = [
  { value: 1, label: 'First option' },
  { value: 2, label: 'Second option' },
];

describe('<Select />', () => {
  it('Renders', () => {
    const renderedValue = renderer.create(
      <Select />,
    );

    expect(renderedValue).toMatchSnapshot();
  });

  it('Renders without placeholder option', () => {
    const renderedValue = renderer.create(
      <Select showPlaceholder={false} />,
    );

    expect(renderedValue).toMatchSnapshot();
  });

  it('Renders with custom placeholder option', () => {
    const renderedValue = renderer.create(
      <Select placeholder="PICK ONE!!!" />,
    );

    expect(renderedValue).toMatchSnapshot();
  });

  it('Renders with options', () => {
    const renderedValue = renderer.create(
      <Select options={options} />,
    );

    expect(renderedValue).toMatchSnapshot();
  });

  it('Renders with name prop', () => {
    const renderedValue = renderer.create(
      <Select name="theSelectizer" />,
    );

    expect(renderedValue).toMatchSnapshot();
  });

  it('Renders with value prop', () => {
    const renderedValue = renderer.create(
      <Select value={1} />,
    );

    expect(renderedValue).toMatchSnapshot();
  });

  it('Applies className prop to class attribute', () => {
    const item = shallow(
      <Select className="added" />,
    );

    expect(item.find('.select.added').exists()).toEqual(true);
  });

  it('Applies optionClassName prop to child options class attributes', () => {
    const wrapper = render(
      <Select
        options={options}
        optionClassName="added" />,
    );

    expect(wrapper.find('.option.added').length).toEqual(3);
  });

  it('simulates change events', () => {
    const onChangeMock = jest.fn();
    const event = { target: { value: 2 } };

    const item = shallow(
      <Select
        value={1}
        options={options}
        onChange={onChangeMock} />,
    );

    item.find('select').simulate('change', event);

    expect(onChangeMock.mock.calls.length).toBe(1);
  });
});
