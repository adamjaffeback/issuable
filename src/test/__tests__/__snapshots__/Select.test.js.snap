// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> Renders 1`] = `
<div
  className="selectWrapper"
>
  <select
    className="select"
    onChange={[Function]}
    value={Object {}}
  >
    <option
      className="option"
      hidden={true}
      value=""
    >
      Select
    </option>
  </select>
</div>
`;

exports[`<Select /> Renders with custom placeholder option 1`] = `
<div
  className="selectWrapper"
>
  <select
    className="select"
    onChange={[Function]}
    value={Object {}}
  >
    <option
      className="option"
      hidden={true}
      value=""
    >
      PICK ONE!!!
    </option>
  </select>
</div>
`;

exports[`<Select /> Renders with name prop 1`] = `
<div
  className="selectWrapper"
>
  <select
    className="select"
    name="theSelectizer"
    onChange={[Function]}
    value={Object {}}
  >
    <option
      className="option"
      hidden={true}
      value=""
    >
      Select
    </option>
  </select>
</div>
`;

exports[`<Select /> Renders with options 1`] = `
<div
  className="selectWrapper"
>
  <select
    className="select"
    onChange={[Function]}
    value={Object {}}
  >
    <option
      className="option"
      hidden={true}
      value=""
    >
      Select
    </option>
    <option
      className="option"
      value={1}
    >
      First option
    </option>
    <option
      className="option"
      value={2}
    >
      Second option
    </option>
  </select>
</div>
`;

exports[`<Select /> Renders with value prop 1`] = `
<div
  className="selectWrapper"
>
  <select
    className="select"
    onChange={[Function]}
    value={1}
  >
    <option
      className="option"
      hidden={true}
      value=""
    >
      Select
    </option>
  </select>
</div>
`;

exports[`<Select /> Renders without placeholder option 1`] = `
<div
  className="selectWrapper"
>
  <select
    className="select"
    onChange={[Function]}
    value={Object {}}
  />
</div>
`;
